crates/ir-generated/src/emit/emitter.rs

```diff
@@ -0,0 +1,87 @@
+use std::mem;
+
+use rokugo_ir::{
+    container::IrContainer,
+    op_code::IrOpCode,
+    register::{chill::RegisterChill, natural::RegisterNat32, Register, RegisterId},
+};
+
+use crate::errors::register::RegisterAllocationError;
+
+use super::register_allocator::{RegisterAllocator, RegisterDropGuard};
+
+pub struct IrEmitter {
+    data: Vec<u8>,
+    register_allocator: RegisterAllocator,
+}
+
+impl IrEmitter {
+    pub fn new() -> Self {
+        IrEmitter {
+            data: Vec::new(),
+            register_allocator: RegisterAllocator::new(),
+        }
+    }
+}
+
+/// # Local Memory
+impl IrEmitter {
+    /// Allocates a virtual register, or prepare a native register to store a new 32-bit natural value.
+    pub fn alloc_register_nat32(
+        &mut self,
+        chill: RegisterChill,
+    ) -> Result<RegisterDropGuard<RegisterNat32>, RegisterAllocationError> {
+        let id = self.register_allocator.next_nat32()?;
+        self.emit(IrOpCode::AllocRegisterNat32);
+        self.emit_register_id(id);
+        self.data.extend_from_slice(&chill.to_le_bytes());
+
+        // SAFETY: This is safe, because `RegisterNat32` is a wrapper around `RegisterId`.
+        Ok(RegisterDropGuard::new(unsafe { mem::transmute(id) }))
+    }
+
+    /// Drops a virtual register, what preverts it from being chilled.
+    /// # Remarks
+    /// This instruction is not dropping any memory like pointer etc. It is only a hint to the JIT compiler.
+    pub fn drop_register<T: Register>(&mut self, register: RegisterDropGuard<T>) -> &mut Self {
```

Comment:
I'm of the opinion that for functions, we should be preferring `where` clauses over specifying constraints directly in generic arguments, as constraints tend to make the first line of the signature quite long - and the first line is generally the most important line, as it gives you a brief idea on how the shape of the API looks.

If you don't care about having the `T` argument though and would just like _any_ `RegisterDropGuard<T>`, I believe using `impl T` might be even better:
```suggestion
    pub fn drop_register(&mut self, register: RegisterDropGuard<impl Register>) -> &mut Self {
```